packages/messages/src/browser/notifications-manager.ts

```diff
@@ -0,0 +1,285 @@
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { Md5 } from 'ts-md5';
+import * as markdownit from 'markdown-it';
+import throttle = require('lodash.throttle');
+import { NotificationPreferences } from './notification-preferences';
+import { ContextKeyService, ContextKey } from '@theia/core/lib/browser/context-key-service';
+import { OpenerService } from '@theia/core/lib/browser';
+import URI from '@theia/core/lib/common/uri';
+
+export const NotificationManager = Symbol('NotificationManager');
+export interface NotificationManager {
+    readonly open: boolean;
+    readonly onUpdate: Event<NotificationManager.UpdateEvent>;
+    accept(notification: Notification | string, action: string): void;
+    hide(): void;
+    toggle(): void;
+    clear(notification: Notification | string): void;
+    clearAll(): void;
+    toggleExpansion(notification: string): void;
+    openLink(link: string): Promise<void>;
+}
+export namespace NotificationManager {
+    export interface UpdateEvent {
+        readonly notifications: Notification[];
+        readonly open: boolean;
+    }
+}
+
+export interface Notification {
+    messageId: string;
+    location: Notification.Location;
+    message: string;
+    source?: string;
+    expandable: boolean;
+    collapsed: boolean;
+    type: Notification.Type;
+    actions: string[];
+    progress?: number;
+}
+export namespace Notification {
+    export type Type = 'info' | 'warning' | 'error' | 'progress';
+    export type Location = 'window' | 'notification';
+}
+
+@injectable()
+export class NotificationManagerImpl extends MessageClient implements NotificationManager {
+
+    @inject(NotificationPreferences)
+    protected readonly preferences: NotificationPreferences;
+
+    @inject(ContextKeyService)
+    protected readonly contextKeyService: ContextKeyService;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    protected readonly onUpdateEmitter = new Emitter<NotificationManager.UpdateEvent>();
+    protected readonly fireUpdateEvent = throttle(() => {
+        const notifications = deepClone(Array.from(this.notifications.values()));
+        this.onUpdateEmitter.fire({ notifications, open: this.open });
+    }, 250, { leading: false });
+    readonly onUpdate = this.onUpdateEmitter.event;
+
+    protected readonly resultPromises = new Map<string, Deferred<string | undefined>>();
+    protected readonly notifications = new Map<string, Notification>();
+
+    protected notificationToastsVisibleKey: ContextKey<boolean>;
+    protected notificationCenterVisibleKey: ContextKey<boolean>;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.notificationToastsVisibleKey = this.contextKeyService.createKey<boolean>('notificationToastsVisible', false);
+        this.notificationCenterVisibleKey = this.contextKeyService.createKey<boolean>('notificationCenterVisible', false);
+    }
+    protected updaetContextKeys(): void {
```

Comment:
it is never called, how do you test context keys? Should we remove them from this PR and tackle with another issue/PR?

Reply:
@AlexTugarev please get rid of it from this PR and file an issue to implement as a follow-up for VS Code compatibility

Reply:
should be called now whenever the internal state changes. ✔️